network/p2p/tracer/gossipSubMeshTracer.go

```diff
@@ -0,0 +1,184 @@
+
+	if _, ok := t.topicMeshMap[topic]; !ok {
+		t.topicMeshMap[topic] = make(map[peer.ID]struct{})
+	}
+	t.topicMeshMap[topic][p] = struct{}{}
+
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("grafted peer not found in identity provider")
+		return
+	}
+
+	lg.Info().Hex("flow_id", logging.ID(id.NodeID)).Str("role", id.Role.String()).Msg("grafted peer")
+	t.metrics.OnLocalMeshSizeUpdated(topic, len(t.topicMeshMap[topic]))
+}
+
+// Prune is called when a peer is removed from a topic mesh. The tracer uses this to track the mesh peers.
+func (t *GossipSubMeshTracer) Prune(p peer.ID, topic string) {
+	t.topicMeshMu.Lock()
+	defer t.topicMeshMu.Unlock()
+
+	lg := t.logger.With().Str("topic", topic).Str("peer_id", p.String()).Logger()
+
+	if _, ok := t.topicMeshMap[topic]; !ok {
+		return
+	}
+	delete(t.topicMeshMap[topic], p)
+
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("pruned peer not found in identity provider")
+		return
+	}
+
+	lg.Info().Hex("flow_id", logging.ID(id.NodeID)).Str("role", id.Role.String()).Msg("pruned peer")
+	t.metrics.OnLocalMeshSizeUpdated(topic, len(t.topicMeshMap[topic]))
+}
+
+// logLoop logs the mesh peers of the local node for each topic at a regular interval.
+func (t *GossipSubMeshTracer) logLoop(ctx irrecoverable.SignalerContext) {
+	ticker := time.NewTicker(t.loggerInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			t.logPeers()
+		}
+	}
+}
+
+// logPeers logs the mesh peers of the local node for each topic.
+// Note that based on GossipSub parameters, we expect to have between 6 and 12 peers in the mesh for each topic.
+// Hence, choosing a heartbeat interval in the order of minutes should be sufficient to log the mesh peers of the local node.
+// Also, note that the mesh peers are also logged reactively when a peer is added or removed from the mesh.
+func (t *GossipSubMeshTracer) logPeers() {
+	t.topicMeshMu.RLock()
+	defer t.topicMeshMu.RUnlock()
+
+	for topic := range t.topicMeshMap {
+		shouldWarn := false // whether we should warn about the mesh state
+		lg := t.logger.With().Dur("heartbeat_interval", t.loggerInterval).Str("topic", topic).Logger()
+		for p := range t.topicMeshMap[topic] {
+			id, exists := t.idProvider.ByPeerID(p)
+			if !exists {
+				shouldWarn = true
+				lg = lg.With().
+					Str("peer_id", p.String()).
+					Str("flow_id", "unknown").
+					Str("role", "unknown").
+					Logger()
+
+				continue
+			}
+
+			lg = lg.With().
+				Str("peer_id", p.String()).
+				Hex("flow_id", logging.ID(id.NodeID)).
+				Str("role", id.Role.String()).
+				Logger()
+		}
+
+		if shouldWarn {
+			lg.Warn().
+				Bool(logging.KeySuspicious, true).
+				Msg(MeshLogIntervalMsg)
+			continue
+		}
+		lg.Info().Msg(MeshLogIntervalMsg)
```

Comment:
A couple of questions here:
1. Does this need to be logged at `Info` level, or could it be `Debug`/`Trace`? Depending on how often `logPeers` is called, this could add quite a bit of data to the logs.
2. Does it make sense to just log counts here, and move the data dump to an admin command?

Reply:
> Does this need to be logged at Info level, or could it be Debug/Trace? Depending on how often logPeers is called, this could add quite a bit of data to the logs.

To facilitate debugging and incident response, it's helpful to have topology logs available at the Info level so that we can also have insight into the recent topology updates if any. Currently, these logs are being generated at 1-minute intervals. However, if it's found to be too noisy, we can adjust the interval to be longer, such as every 5 minutes. The logging interval can be configured by setting a flag.

> Does it make sense to just log counts here, and move the data dump to an admin command?

The admin command for topology monitoring may not be the most efficient approach for topology-related incidents. In order to have proper observability of the recent change history of the topology, as well as any real-time changes that happen to it, we need a more comprehensive solution. For instance, we may need to see whether the mesh of a particular topic had a uniform distribution of participating roles in recent history. The admin command, unfortunately, only provides a snapshot (as well as the forward history).

To overcome this limitation, the current approach provides insights into the entire topology updates of the nodes through our single logging interface. This will avoid the need to SSH into each node and run admin commands individually, which can be time-consuming and inefficient during the incidents. With a logging dashboard, we can quickly identify any anomalies or inconsistencies in the topology and take corrective actions accordingly.